Mathlib/Computability/TuringDegrees.lean

```diff
@@ -0,0 +1,298 @@
+  | zero : RecursiveIn O (fun _ => 0)
+  | succ : RecursiveIn O Nat.succ
+  | left : RecursiveIn O (fun n => (Nat.unpair n).1)
+  | right : RecursiveIn O (fun n => (Nat.unpair n).2)
+  | oracle : ∀ g ∈ O, RecursiveIn O g
+  | pair {f h : ℕ →. ℕ} (hf : RecursiveIn O f) (hh : RecursiveIn O h) :
+      RecursiveIn O (fun n => (Nat.pair <$> f n <*> h n))
+  | comp {f h : ℕ →. ℕ} (hf : RecursiveIn O f) (hh : RecursiveIn O h) :
+      RecursiveIn O (fun n => h n >>= f)
+  | prec {f h : ℕ →. ℕ} (hf : RecursiveIn O f) (hh : RecursiveIn O h) :
+      RecursiveIn O (fun p =>
+        let (a, n) := Nat.unpair p
+        n.rec (f a) (fun y IH => do
+          let i ← IH
+          h (Nat.pair a (Nat.pair y i))
+        )
+      )
+  | rfind {f : ℕ →. ℕ} (hf : RecursiveIn O f) :
+      RecursiveIn O (fun a =>
+        Nat.rfind (fun n => (fun m => m = 0) <$> f (Nat.pair a n))
+      )
+/--
+f is turing reducible to g if f is recursive in g
+-/
+abbrev TuringReducible (f g : ℕ →. ℕ) : Prop :=
+  RecursiveIn {g} f
+
+/--
+Custom infix notation for `TuringReducible`.
+-/
+infix:50 "≤ᵀ" => TuringReducible
+
+/--
+`f` is Turing equivalent to `g` if `f` is reducible to `g` and `g` is reducible to `f`.
+-/
+abbrev TuringEquivalent (f g : ℕ →. ℕ) : Prop :=
+  AntisymmRel TuringReducible f g
+
+/--
+Custom infix notation for `TuringEquivalent`.
+-/
+infix:50 "≡ᵀ" => TuringEquivalent
+
+/--
+If a function is partial recursive, then it is recursive in every partial function.
+-/
+lemma Nat.Partrec.turingReducible (f : ℕ →. ℕ) (pF : Nat.Partrec f) (g : ℕ →. ℕ) : f ≤ᵀ g := by
+  induction pF
+  · exact RecursiveIn.zero
+  · exact RecursiveIn.succ
+  · exact RecursiveIn.left
+  · exact RecursiveIn.right
+  · case pair f' g' _ _ ih1 ih2 =>
+    apply RecursiveIn.pair ih1 ih2
+  · case comp f' g' _ _ ih1 ih2 =>
+    apply RecursiveIn.comp ih1 ih2
+  · case prec f' g' _ _ ih1 ih2 =>
+    apply RecursiveIn.prec ih1 ih2
+  · case rfind f' _ ih =>
+    apply RecursiveIn.rfind ih
+
+/--
+If a function is recursive in the constant zero function,
+then it is partial recursive.
+-/
+lemma TuringReducible.partrec_of_zero (f : ℕ →. ℕ) (fRecInZero : f ≤ᵀ fun _ => Part.some 0) :
+  Nat.Partrec f := by
+  induction fRecInZero
+  · exact Nat.Partrec.zero
+  · exact Nat.Partrec.succ
+  · exact Nat.Partrec.left
+  · exact Nat.Partrec.right
+  · case oracle g hg =>
+    simp at hg; rw [hg]
+    exact Nat.Partrec.zero
+  · case pair g h _ _ ih1 ih2 =>
+    exact Nat.Partrec.pair ih1 ih2
+  · case comp g h _ _ ih1 ih2 =>
+    exact Nat.Partrec.comp ih1 ih2
+  · case prec g h _ _ ih1 ih2 =>
+    exact Nat.Partrec.prec ih1 ih2
+  · case rfind g _ ih =>
+    exact Nat.Partrec.rfind ih
+
+/--
+A partial function `f` is partial recursive if and only if it is recursive in
+every partial function `g`.
+-/
+theorem partrec_iff_partrec_in_everything (f : ℕ →. ℕ) : Nat.Partrec f ↔ ∀ g, f ≤ᵀ g :=
+  ⟨(·.turingReducible), (· _ |>.partrec_of_zero)⟩
+
+/--
```

Comment:
You don't neeed to attach doc comments to theorems and instances, unless they are not self-explanitory. However, all `def`s need doc comments.